packages/gatsby/src/schema/resolvers.ts

```diff
@@ -221,20 +224,42 @@ export function link<TSource, TArgs>(
     TArgs
   >
 ): GatsbyResolver<TSource, TArgs> {
-  return async function linkResolver(
+  // Note: we explicitly make an attempt to prevent using the `async` keyword because often
+  //       it does not return a promise and this makes a significant difference at scale.
+
+  return function linkResolver(
     source,
     args,
     context,
     info
-  ): Promise<any> {
+  ): ResolvedLink | Promise<ResolvedLink> {
     const resolver = fieldConfig.resolve || context.defaultFieldResolver
-    const fieldValue = await resolver(source, args, context, {
+    const fieldValueOrPromise = resolver(source, args, context, {
       ...info,
       from: options.from || info.from,
       fromNode: options.from ? options.fromNode : info.fromNode,
     })
 
-    if (fieldValue == null) return null
+    // Note: for this function, at scale, conditional .then is more efficient than generic await
+    if (fieldValueOrPromise?.then) {
```

Comment:
I think we should still do `typeof fieldValueOrPromise?.then === "function"` otherwise it has the potential to break for some people (e.g. with a value like `{ then: '2012-12-12' }` - highly unlikely but still possible).

Also, graphql-js uses the exact same check here: https://github.com/graphql/graphql-js/blob/master/src/jsutils/isPromise.js

```suggestion
    if (typeof fieldValueOrPromise?.then === `function`) {
```

Reply:
Ok sounds good to me